### PR TITLE
Docker build json

### PR DESCRIPTION
**Pull request by agnihotrigaurav**
User can specify map of registry with username and password as per below snippet in json format.
```
internal/docker-build:
  dockerfile: Dockerfile
  image-name: myImage
  registry-auth-config: |
    {
      "docker.example.com": {
        "username": "janedoe",
        "password": "hunter2"
    },
      "https://index.docker.io/v1/": {
        "username": "mobydock",
        "password": "conta1n3rize14"
    }
  }
```
Integration test case not added as requires username and password.

Staging test are as follows.

Base image is from private repository and requires authentication.(result=PASS)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4d870ee8a21100276de2d0

Authentication not required. Base image is from public repo(result=PASS)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4d79bbe8a21100276de287

When username or password is incorrect and base image requires authentication(result=failed).
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4d891bb6fbc800067498b2

Authentication details not provided for private repo(result=failed)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b45b0a93893340001aa29e5

Incorrect Registry specified(result=failed).
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4d8983b6fbc800067498b9

### CODE REVIEW

File: docker/docker_build.go [hunk 1]
```diff
@@ -150,11 +153,51 @@ func (s *DockerBuildStep) configure(env *util.Environment) {
...
 	}
 
+	if registryAuthConfig, ok := s.data["registry-auth-config"]; ok {
+		in := []byte(registryAuthConfig)
+		var raw map[string]interface{}
+		err := json.Unmarshal(in, &raw)
+		if err != nil {
+			return err
+		}
+		authconfigs := make(map[string]types.AuthConfig)
```
[the remark this answers is not in the document]
Comment (agnihotrigaurav) [reply docker/docker_build.go]: Do you think any performance benefit...If we do that...we need to loop again for interpolation

**Comment by agnihotrigaurav** [comment]
Implemented above comment and tested below scenarios again on staging.

01) Base image is from private repository and requires authentication.(result=PASS)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4f08f8e8a21100276de928

================================================================== 
02) Authentication not required. Base image is from public repo(result=PASS)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4f0bf9e8a21100276de94e

================================================================== 
03) When username or password is incorrect and base image requires authentication(result=failed).
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4f09c8e8a21100276de92f?step=5b4f09dc3ff5480007e986ed

================================================================== 
04) Authentication details not provided for private Image(result=failed)
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4f0b7ce8a21100276de93f?step=5b4f0b903ff5480007e9875d

================================================================== 
05) Incorrect Registry specified(result=failed).
https://dev.wercker.com/gauagnih/wercker-docker-build/runs/build/5b4f0b0cb6fbc80006749b0b?step=5b4f0b1f3ff5480007e9870c

File: docker/docker_build.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (agnihotrigaurav) [reply docker/docker_build.go]: Done.

**Comment by hatchan** [comment]
I don't understand why we're not using simple username/password properties?

**Comment by hatchan** [comment]
Instead of passing in a JSON blob, can't we create a array with auth objects?

Example:

```
internal/docker-build:
  dockerfile: Dockerfile
  image-name: myImage
  registry-auth-config: |
    {
      "docker.example.com": {
        "username": "janedoe",
        "password": "hunter2"
      },
      "https://index.docker.io/v1/": {
        "username": "mobydock",
        "password": "conta1n3rize14"
      }
    }
```

Would be:

```
internal/docker-build:
  dockerfile: Dockerfile
  image-name: myImage
  auth:
  - registry: docker.example.com
    username: janedoe
    password: hunter2
  - username: mobydock
    password: conta1n3rize14
```
(omitting the registry, defaults to Docker Hub)

**Comment by gilbode** [comment]
@hatchan - i believe there were other suggestions along the lines of changing the type of the parameter to something other than string.  my feedback was that i think we should think about this more holistically if we're going to extend the syntax of step parameters vs. just doing that for this particular case.  particularly for non-internal steps it was unclear how we would pass the value of such a structure to the step.

**Pull request merged**